lib/strings.nix

```diff
@@ -1320,4 +1355,44 @@ rec {
         else if k == 2 then infixDifferAtMost2 ainfix binfix
         else levenshtein ainfix binfix <= k;
       in f;
+
+  /*
+    Return the derivation path that has the given path as an output.
+
+    Not all paths are constructed by derivations, so this function may throw.
+
+    Design note:
+    While this function is close to being the inverse of the RFC 92 `outputOf`
+    function, it does not return the output name. This is intentional, because
+    the output name *of a derivation* is not meant to carry meaning, unlike
+    the output *attribute* of a package. Hence, bundling the output name into
+    this return value would almost always be inconvenient.
+  */
+  derivationOf =
+    v:
+    let
+      outputString =
+        if isDerivation v
+        then "${v.outPath}"
+        else if isString v
+        then v
+        else if isPath v
+        then throw "derivationOf: The passed argument is a path value, and path values are not outputs, so we can't find the derivation that produced it."
+        else throw "derivationOf: Expected a string containing a derivation output, but got a ${typeOf v} instead.";
+      ctx = builtins.getContext outputString;
+      items = attrsToList ctx;
+      drv = (head items).name;
+      group = (head items).value;
+      drvAsSource = builtins.appendContext drv { ${drv} = { path = true; }; };
+    in
+      throwIf (! hasPrefix builtins.storeDir outputString)
+        "lib.derivationOf: To retrieve the derivation of an output, the passed string must reference a store path. The passed string does not reference a store path."
```

Comment:
Probably better to not just check for a store dir prefix, but rather that the entire string is a store path. Or is there a case where that wouldn't work?

Reply:
It would break `derivationOf (getExe foo)`.

We'd not be able to let `writeScript` return a proper package with a `bin` directory, e.g. `let script = mkDerivation ...; in script // { outPath = lib.getExe script; }`, and then pass that to `derivationOf`.

Reply:
What's the use case for this?

Reply:
A problem with derivations that output a single file is that they can not be updated to contain anything else, such as a `nix-support` directory, or auxiliary files that need to reference the original thing. These would be breaking changes.

However, if functions such as `writeScript` could return a subpath in a store path, that frees up the store path for such new files in the future.

It also solves this annoyance:
- https://github.com/NixOS/nixpkgs/issues/147769